### ISSUE TITLE
RT crashes when opens a folder in file browser and then delete it in Windows explorer

### ISSUE THREAD

**Issue by syyrmb** [Beep6581/RawTherapee#6176]
**Short description**
When opening a folder in RawTherapee's file browser, and then delete that folder in Windows 10s' explorer, RawTherapee will freeze up and requires you to force-quit it.

**Steps to reproduce**
1. Open RawTherapee and open a folder (double click on it).
2. Delete that folder in Windows explorer.
3. Back to RawTherapee and click on its interface.
4. Freeze (RawTherapee not responding)

**Expected behavior**
RawTherapee should updates its file browser and removes the deleted folder without freezing.

**Additional information**
- This bug occurs in the public release version of RawTherapee 5.8, and RT dev. 5.8-2776-gf72747771_20210224 suffers from it too.
- OS: Windows 10 2004

Other useful information:
- I tried to replicate this bug on my elementary os with RT 5.8 public release, turns out it seems like the bug can't be carried out on Linux.

**Comment by MJacal** [comment]
I can not reproduce this with RawTherapee_dev_5.8-2822-g170df0393_20210320 on Windows 10, RT does not crash, no problems of any kind here.

**Comment by syyrmb** [comment]
> 
> 
> I can not reproduce this with RawTherapee_dev_5.8-2822-g170df0393_20210320 on Windows 10, RT does not crash, no problems of any kind here.


Thanks for testing. That's pretty weird, then (Maybe it's my computer being weird again). I recorded a video demo, hopefully it can be useful (I blurred some parts due to privacy concerns).
[RT_crash_demo.zip](https://github.com/Beep6581/RawTherapee/files/6181862/RT_crash_demo.zip)

**Comment by MJacal** [comment]
Update: I can reproduce it, but only when the test folder, that gets deleted, is empty. In my first attempt I have put a couple of photos into it.

Same behaviour on my Windows 10 laptop with RT 5.8-2461.

**Comment by heckflosse** [comment]
I will try to reproduce this...

**Comment by heckflosse** [comment]
I created an empty folder at my C: drive `C:\empty`, opened the folder in RawTherapee, switched to Windows explorer, deleted the folder, switched to RawTherapee and got no crash. I must miss something...

**Comment by heckflosse** [comment]
@syyrmb Your video shows a freeze, not a crash. Though in my tests I even get no freeze...

**Comment by syyrmb** [comment]
> 
> 
> @syyrmb Your video shows a freeze, not a crash. Though in my tests I even get no freeze...

Yeah this is not a crash, I'll fix that. If you don't double click on the "empty" folder in RT, then it'll be fine when you delete it.

**Comment by heckflosse** [comment]
I double clicked in RT on the non existing (after deletion) empty folder and still no freeze or crash

**Comment by syyrmb** [comment]
> 
> 
> I double clicked in RT on the non existing (after deletion) empty folder and still no freeze or crash

Sorry for not being clear, the first step is double clicking on the empty folder, then delete it, and finally going back to RT.

**Comment by heckflosse** [comment]
> Sorry for not being clear, the first step is double clicking on the empty folder, then delete it, and finally going back to RT.

That was quite clear and exactly what I did without a crash or freeze. Really strange...

**Comment by Thanatomanic** [comment]
I can confirm the freeze...

Edit: for reference (the file is sent to the recycle bin, but the dialog is on my second screen)

https://user-images.githubusercontent.com/6567747/112000949-bfea1180-8b1e-11eb-888d-522e56d00d1e.mp4

**Comment by syyrmb** [comment]
> 
> 
> > Sorry for not being clear, the first step is double clicking on the empty folder, then delete it, and finally going back to RT.
> 
> That was quite clear and exactly what I did without a crash or freeze. Really strange...

Where did you deleted folder go, to the recycle bin or totally deleted? I found that using shift+delete is fine.

**Comment by heckflosse** [comment]
@Thanatomanic 

> I can confirm the freeze...

Can you nail it with a debug build?

**Comment by heckflosse** [comment]
@syyrmb 

> Where did you deleted folder go, to the recycle bin or totally deleted? I found that using shift+delete is fine.

Good point. I totally deleted in my test...

**Comment by Thanatomanic** [comment]
Okay, so far:
1. Confirmed that removing an empty folder that is also open in RT's file browser, freezes the program
2. Confirmed that shift-deleting the empty folder has no problem.
3. Confirmed that removing a non-empty folder does not freeze RT, however...
4. Dubious behavior: removing a non-empty folder removes the folder reference in the "Folders" list, but the actual file browser portion stays unchanged. I would assume this gets emptied, since the files do not exist anymore. Trying to open an image results in a "Cannot load image" message, and luckily does not freeze the program. However, what does this all mean for the cached data?

Anyway, I'll try @heckflosse 's suggestion for a debug build!

**Comment by heckflosse** [comment]
@Thanatomanic 

 > Anyway, I'll try @heckflosse 's suggestion for a debug build!

Keep in mind that debugging a frozen program is not straight forward using gdb on msys2. You can not use CTRL-C because then gdb will terminate. You will have to use another msys2 terminal and attach gdb to the process of the frozen RT session. There is an issue somewhere in RT Github, but I didn't find it on the rush...

**Comment by syyrmb** [comment]
I was messing around with RT 5.13 dev build and accidentally found out that deleting a folder no longer results in a freezing. Can we close this issue now if we can confirm this bug no longer happens?

**Comment by Lawrence37** [comment]
I will close this. If someone is able to reproduce it again, we can reopen it.
